src/main/java/appeng/me/cluster/implementations/CraftingCPUCluster.java

```diff
@@ -397,6 +400,23 @@ private void completeJob() {
             AELog.crafting(LOG_MARK_AS_COMPLETE, logStack);
         }
 
+        // See https://github.com/AppliedEnergistics/Applied-Energistics-2/issues/5158
+        // We suspect this is caused by waitingFor not being empty when the job is marked as complete
+        // This code is intended to analyze this problem further if it occurs again
+        if (!this.waitingFor.isEmpty()) {
+            StringBuilder errorMessage = new StringBuilder("A crafting job completed while still waiting for items:\n");
+            for (IAEItemStack stack : this.waitingFor) {
+                errorMessage.append(" - ").append(stack).append("\n");
+            }
+            errorMessage.append("Current inventory:\n");
+            for (IAEItemStack stack : inventory.getItemList()) {
+                errorMessage.append(" - ").append(stack).append("\n");
+            }
+            errorMessage.append("Final output: ").append(finalOutput.copy().setStackSize(this.startItemCount));
+            throw new RuntimeException(errorMessage.toString());
+        }
+        this.waitingFor.resetStatus();
```

Comment:
Wouldn't that be actually dead code/useless?

Reply:
It's for "the future" when we actually remove the fail-fast code. I do want to clear that list :-P

Reply:
It's still wrong. A crafting job can never be complete, when it's still waiting on more items. So it's always reseting an empty list, which is a bit pointless. Or it's actually an unfixed bug.

Reply:
It still shows what the invariant is. We _can_ also replace this with Precondtion.checkState(waitingFor.isEmpty()) if you prefer that. I just don't  want this to be forgotten when we do remove the in-depth check .